sdk/include/ds/ring_buffer.h

```diff
@@ -0,0 +1,171 @@
+// Copyright Microsoft and CHERIoT Contributors.
+// SPDX-License-Identifier: MIT
+
+/**
+ * @file Ring/circular buffer state machine.
+ */
+
+#pragma once
+
+#include <cdefs.h>
+#include <type_traits>
+
+namespace ds::ring_buffer
+{
+	/**
+	 * A statically-sized, non-atomic ring buffer state machine.  The actual
+	 * element storage must be provided externally; all this does is manage the
+	 * cursors.  This representation uses an explicit empty flag, rather than
+	 * considering equal cursors to be empty, so as to not need an extra element
+	 * of store Capacity-many items.
+	 *
+	 * The type of the cursors is parametric, but must be unsigned.
+	 *
+	 * The consumer workflow is:
+	 *   - use head_get() to test emptiness and, if nonempty, read the index
+	 *     of the head element
+	 *   - make use of the element in the (external!) storage at that index
+	 *   - call head_advance() to discard the head element
+	 *
+	 * The producer workflow is:
+	 *   - use tail_next() to retrieve the index for the next element,
+	 *     if there is room
+	 *   - populate the element in the (external!) storage at that index
+	 *   - call tail_advance() to make that element available to the
+	 *     consumer
+	 */
+	template<typename Debug, std::size_t Capacity, typename _Ix = std::size_t>
+	class Cursors
+	{
+		static_assert(std::is_arithmetic_v<_Ix> && std::is_unsigned_v<_Ix>,
+		              "Ring buffer cursor type must be unsigned arithmetic");
+
+		template<typename... Args>
+		using Assert = typename Debug::template Assert<Args...>;
+
+		public:
+		using Ix = _Ix;
+
+		private:
+		Ix   head;
+		Ix   tail;
+		bool empty;
+
+		Ix advance(Ix v)
+		{
+			if constexpr ((Capacity & (Capacity - 1)) == 0)
+			{
+				/* For power of two sizes, we can rely on bit masking. */
+				return (v - 1) & (Capacity - 1);
+			}
+			else
+			{
+				/*
+				 * For non-power-of-two, we can use unsigned underflow and
+				 * signed arithmetic right shift to avoid branching.  mask is
+				 * all zeros if next is non-negative and all ones otherwise.
+				 */
+				Ix next   = v - 1;
+				using SIx = std::make_signed_t<Ix>;
+				Ix mask   = static_cast<SIx>(next) >> (8 * sizeof(SIx) - 1);
+				return (mask & (Capacity - 1)) | (~mask & next);
```

Comment:
It's is unclear to me if this is fast than
```C++
return (v == 0 ? Capacity : v) - 1
```
If there a conditional move on this platform?

Reply:
I think we do not have a conditional move, no, but this might still be slower than the alternative.